alibi_detect/utils/pytorch/kernels.py

```diff
@@ -14,22 +18,76 @@ def __init__(self, sigma: Optional[torch.Tensor] = None) -> None:
         Parameters
         ----------
         sigma
-            Optional sigma used for the kernel.
+            Bandwidth used for the kernel. Needn't be specified if being inferred or trained.
+            Can pass multiple values to eval kernel with and then average.
+        trainable
+            Whether or not to track gradients w.r.t. sigma to allow it to be trained.
         """
         super().__init__()
-        self.sigma = sigma
+        if sigma is None:
+            self.log_sigma = nn.Parameter(torch.empty(1), requires_grad=trainable)
+            self.init_required = True  # TODO: Is this right now?
```

Comment:
Looks right, although the ValueError in the forward pass could already have been thrown at init if `trainable=True` and `sigma=None`. One alternative would be to infer a sigma first via the median heuristic and set trainable True afterwards for log_sigma. But that might be more of a non-priority edge case.

Reply:
Sorry I'm unsure on both your points here.

On first point -- having `trainable=True` and `sigma=None` is the default for a trainable `GaussianRBF` and we wouldn't want it to lead to an error?

On second point -- I don't think we'd want that to throw an error either? The trainable `GaussianRBF` is always called with `infer_sigma=False`.  We just temporarily need `init_required=True` to override this for the first call.

Reply:
Yes it became clear when it is used in the spot the difference detector. So all good and fine to remove the TODO.